packages/serverpod_test/lib/src/test_session.dart

```diff
@@ -0,0 +1,142 @@
+          {String? authId}) =>
+      _AuthenticationInfoOverride(userId, scopes, authId: authId);
+}
+
+/// Overrides the authenticationInfo on the session. This will bypass any auth handlers.
+class _AuthenticationInfoOverride extends AuthenticationOverride {
+  final AuthenticationInfo _authenticationInfo;
+
+  /// The authentication info to use for the session.
+  AuthenticationInfo get authenticationInfo => _authenticationInfo;
+
+  /// Creates a new AuthenticationInfoOverride with the provided authentication info.
+  _AuthenticationInfoOverride(int userId, Set<Scope> scopes, {String? authId})
+      : _authenticationInfo =
+            AuthenticationInfo(userId, scopes, authId: authId);
+}
+
+/// Does not override the authentication state in the session.
+class _Unauthenticated extends AuthenticationOverride {}
+
+/// A test specific session that is used to call database methods or pass to endpoints.
+/// This is the public interface exposed to developers.
+abstract class TestSession implements DatabaseAccessor {
+  /// AuthenticationInfo for the session.
+  Future<AuthenticationInfo?> get authenticationInfo;
+
+  /// Creates a new unique session with the provided properties.
+  /// This is useful for setting up different session states in the tests
+  /// or simulating multiple users.
+  TestSession copyWith({
+    AuthenticationOverride? authentication,
+    bool? enableLogging,
+  });
+}
+
+/// Internal test session that contains implementation details that should
+/// only be used internally by the test tools.
+class InternalTestSession extends TestSession {
+  final List<InternalTestSession> _allTestSessions;
+  final TestServerpod _testServerpod;
+  AuthenticationOverride? _authenticationOverride;
+  final bool _enableLogging;
+
+  /// The underlying Serverpod session
+  InternalServerpodSession serverpodSession;
+
+  @override
+  Database get db => serverpodSession.db;
+
+  @override
+  Future<AuthenticationInfo?> get authenticationInfo async {
+    return serverpodSession.authenticated;
+  }
+
+  Transaction? _transaction;
+  @override
+  Transaction? get transaction => _transaction;
+
+  set transaction(Transaction? transaction) {
+    _transaction = transaction;
+    serverpodSession.transaction = transaction;
+  }
+
+  /// Creates a new internal test session.
+  InternalTestSession(
+    TestServerpod testServerpod, {
+    AuthenticationOverride? authenticationOverride,
+    InternalTestSession? sessionWithDatabaseConnection,
+    Transaction? transaction,
+    required bool enableLogging,
+    required List<InternalTestSession> allTestSessions,
+    required this.serverpodSession,
+  })  : _allTestSessions = allTestSessions,
+        _authenticationOverride = authenticationOverride,
+        _testServerpod = testServerpod,
+        _enableLogging = enableLogging,
+        _transaction = transaction {
+    _allTestSessions.add(this);
+    _configureServerpodSession(serverpodSession);
+  }
+
+  @override
+  TestSession copyWith({
```

Comment:
Comment: As discussed offline this name may not exactly be the expected behavior. Not sure if I got a better name in mind, but a more accurate name is something like `createFrom` or similar.

Reply:
Lets discuss together offline!